proving/proving.go

```diff
@@ -237,68 +240,79 @@ func (p *Prover) tryNonces(numLabels uint64, challenge Challenge, startNonce, en
 	var indices []byte
 
 	workersChans := make([]chan []byte, numWorkers)
+	// workersComplete channel will be closed when worker stops listening for appropriate workersChan
+	workersComplete := make([]chan struct{}, numWorkers)
 	for i := range workersChans {
 		workersChans[i] = make(chan []byte, 1024)
+		workersComplete[i] = make(chan struct{})
 	}
-	resultsChan := make(chan nonceResult, numWorkers)
-	errChan := make(chan error, numWorkers)
+	resultsChan := make(chan *nonceResult, numWorkers)
+	errChan := make(chan error, 1)
+
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
 	// Start IO worker.
 	// Feed all labels into each worker chan.
+	wg.Add(1)
 	go func() {
+		defer wg.Done()
 		for {
 			label, err := gsReader.ReadNext()
 			if err != nil {
+				if !errors.Is(err, io.EOF) {
+					errChan <- err
+				}
 				for i := range workersChans {
 					close(workersChans[i])
 				}
-
-				if err != io.EOF {
-					errChan <- err
-				}
-				break
+				return
 			}
 
 			for i := range workersChans {
-				workersChans[i] <- label
+				select {
+				case workersChans[i] <- label:
+				case <-workersComplete[i]:
+				case <-ctx.Done():
+					return
+				}
 			}
 		}
 	}()
 
 	// Start a worker for each nonce.
-	ctx, cancel := context.WithCancel(context.Background())
+	// TODO(dshulyak) it would be more efficient to start a worker per CPU and distribute work among
+	// them but it is not trivial
 	for i := uint32(0); i < numWorkers; i++ {
 		i := i
+		wg.Add(1)
 		go func() {
 			nonce := startNonce + i
 			indices, err = p.tryNonce(ctx, numLabels, challenge, nonce, workersChans[i], difficulty)
-			resultsChan <- nonceResult{nonce, indices, err}
+			close(workersComplete[i])
+			resultsChan <- &nonceResult{nonce, indices, err}
+			wg.Done()
 		}()
 	}
 
 	// Drain the workers results chan.
-	var solutionNonceResult *nonceResult
+	var result *nonceResult
```

Comment:
After your fix for https://github.com/spacemeshos/post/pull/47#discussion_r695368434, it doesn't look like we need this anymore.